src/crystal/system/win32/socket.cr

```diff
@@ -0,0 +1,425 @@
+
+    socket
+  end
+
+  private def initialize_handle(handle)
+    value = 1_u8
+    ret = LibC.setsockopt(handle, LibC::SOL_SOCKET, LibC::SO_REUSEADDR, pointerof(value), 1)
+    if ret == LibC::SOCKET_ERROR
+      raise ::Socket::Error.from_wsa_error("setsockopt")
+    end
+  end
+
+  private def system_connect(addr, timeout = nil)
+    if type.stream?
+      system_connect_stream(addr, timeout) { |error| yield error }
+    else
+      system_connect_connectionless(addr, timeout) { |error| yield error }
+    end
+  end
+
+  private def system_connect_stream(addr, timeout)
+    address = LibC::SockaddrIn6.new
+    address.sin6_family = family
+    address.sin6_port = 0
+    unless LibC.bind(fd, pointerof(address).as(LibC::Sockaddr*), sizeof(LibC::SockaddrIn6)) == 0
+      return yield ::Socket::BindError.from_wsa_error("Could not bind to '*'")
+    end
+
+    error = overlapped_connect(fd, "ConnectEx") do |overlapped|
+      # This is: LibC.ConnectEx(fd, addr, addr.size, nil, 0, nil, overlapped)
+      result = Crystal::System::Socket.connect_ex.call(fd, addr.to_unsafe, addr.size, Pointer(Void).null, 0_u32, Pointer(UInt32).null, overlapped)
+
+      if result.zero?
+        wsa_error = WinError.wsa_value
+
+        case wsa_error
+        when .wsa_io_pending?
+          next
+        when .wsaeaddrnotavail?
+          return yield ::Socket::ConnectError.from_os_error("ConnectEx", wsa_error)
+        else
+          return yield ::Socket::Error.from_os_error("ConnectEx", wsa_error)
+        end
+      end
+    end
+
+    if error
+      yield error
+    end
+  end
+
+  private def system_connect_connectionless(addr, timeout)
+    ret = LibC.connect(fd, addr, addr.size)
+    if ret == LibC::SOCKET_ERROR
+      yield ::Socket::Error.from_wsa_error("connect")
+    end
+  end
+
+  private def system_bind(addr, addrstr)
+    unless LibC.bind(fd, addr, addr.size) == 0
+      yield ::Socket::BindError.from_errno("Could not bind to '#{addrstr}'")
+    end
+  end
+
+  private def system_listen(backlog)
+    unless LibC.listen(fd, backlog) == 0
+      yield ::Socket::Error.from_errno("Listen failed")
+    end
+  end
+
+  protected def system_accept : Handle?
+    client_socket = create_handle(family, type, protocol, blocking)
+    initialize_handle(client_socket)
+
+    if system_accept(client_socket)
+      client_socket
+    else
+      LibC.closesocket(client_socket)
+
+      nil
+    end
+  end
+
+  protected def system_accept(client_socket : Handle) : Bool
+    address_size = sizeof(LibC::SOCKADDR_STORAGE) + 16
+    buffer_size = 0
+    output_buffer = Bytes.new(address_size * 2 + buffer_size)
+
+    success = overlapped_accept(fd, "AcceptEx") do |overlapped|
+      received_bytes = uninitialized UInt32
+
+      result = Crystal::System::Socket.accept_ex.call(fd, client_socket,
+        output_buffer.to_unsafe.as(Void*), buffer_size.to_u32!,
+        address_size.to_u32!, address_size.to_u32!, pointerof(received_bytes), overlapped)
```

Comment:
The value is based on `sizeof` which returns `Int32`, so there must be a conversion at some point. I suppose it could be done earlier, though.
The main reason for this is that the proc call resembles a lib call. The lib call would handle all this type casting (and to_unsafe conversion), but we need to make it explicit here.

Reply:
Fair enough. comment was also about `buffer_size` though.